smart-contracts/wasm-chain-integration/src/v1/mod.rs

```diff
@@ -245,7 +256,11 @@ mod host {
             .checked_add(length)
             .ok_or_else(|| anyhow::anyhow!("Writing past the end of memory."))?
             as usize;
-        let end = std::cmp::min(end, constants::MAX_CONTRACT_STATE as usize) as u32;
+        let end = if limit_return_value_size {
+            std::cmp::min(end, constants::MAX_CONTRACT_STATE as usize) as u32
+        } else {
+            end as u32
```

Comment:
Why are we converting back to u32 here? THat is a bit dangerous although probably impossible to trigger in practice.

I recommend that we explicitly check that `bytes.len()` fits into a u32.

And we don't do conversion from usize to u32 at all.

Reply:
I've tried to clean up all the conversion shenanigans ;)